qiskit/algorithms/linear_solvers/hhl.py

```diff
@@ -0,0 +1,373 @@
+                              observable: Optional[Union[LinearSystemObservable,
+                                                         BaseOperator]] = None,
+                              observable_circuit: Optional[QuantumCircuit] = None,
+                              post_processing: Optional[Callable[[Union[float, List[float]]],
+                                                                 Union[float, List[float]]]] = None,
+                              scaling: float = 1) -> Tuple[Union[float, List[float]],
+                                                           Union[float, List[float]]]:
+        """Calculates the value of the observable(s) given.
+
+        Args:
+            qc: The quantum circuit preparing the solution x to the system.
+            observable: Information to be extracted from the solution.
+            observable_circuit: Circuit to be applied to the solution to extract information.
+            post_processing: Function to compute the value of the observable.
+            scaling: Factor scaling the solution vector.
+
+        Returns:
+            The value of the observable(s) and the circuit results before post-processing as a
+             tuple.
+        """
+        # Get the number of qubits
+        nb = solution.qregs[0].size
+        nl = solution.qregs[1].size
+        na = solution.num_ancillas
+
+        # if the observable is given construct post_processing and observable_circuit
+        if observable is not None:
+            observable_circuit = observable.observable_circuit(nb)
+            post_processing = observable.post_processing
+
+            if isinstance(observable, LinearSystemObservable):
+                observable = observable.observable(nb)
+
+        # in the other case use the identity as observable
+        else:
+            observable = I ^ nb
+
+        # Create the Operators Zero and One
+        zero_op = (I + Z) / 2
+        one_op = (I - Z) / 2
+
+        is_list = True
+        if not isinstance(observable_circuit, list):
+            is_list = False
+            observable_circuit = [observable_circuit]
+            observable = [observable]
+
+        expectations = []
+        for circ, obs in zip(observable_circuit, observable):
+            circuit = QuantumCircuit(solution.num_qubits)
+            circuit.append(solution, circuit.qubits)
+            circuit.append(circ, range(nb))
+
+            ob = one_op ^ TensoredOp((nl + na) * [zero_op]) ^ obs
+            expectations.append(~StateFn(ob) @ StateFn(circuit))
+
+        if is_list:
+            # execute all in a list op to send circuits in batches
+            expectations = ListOp(expectations)
+        else:
+            expectations = expectations[0]
+
+        if self._expectation is not None:
+            expectations = self._expectation.convert(expectations)
+
+        if self._sampler is not None:
+            expectations = self._sampler.convert(expectations)
```

Comment:
So this is like a double conversion if both are supplied (expectation and a quantum instance that results in the sampler being created). The docstring presumably of the two params will say something about this.

Reply:
About the expectation yes, but since the sampler conversion is just the execution of the circuits I don't think we have to explicitly state that. I mean if users pass a quantum instance they'll expect that it's used for circuit execution, right? 🙂

Reply:
The quantum instance is more evident to run the circuits. But self._sampler is set from that -  and that was the only place I saw it used. So despite it being a Quantum Algorithm it seemed the QuantumInstance was optional since the CircuitSample from it could be later bypassed when None. Maybe this is what is confusing me - the expectation object is also optional. I guess we need one or the other or both. Or maybe they are both entirely optional. I am still less than clear from the docstring

Reply:
Added `PauliExpectation()` as the default value if the expectation object is `None`.

Reply:
For VQE the default expectation object is created using the expectation factory that takes into account the backend type. Actually looking at the code it seems this is used too - so the default is one supplied from the Factory, which is not necessarily a PauliExpectation